fdbserver/DDTeamCollection.actor.cpp

```diff
@@ -191,6 +191,93 @@ class DDTeamCollectionImpl {
 		req.reply.send(std::make_pair(res, false));
 	}
 
+	// Returns the overall best team that matches the requirement from `req`. When preferWithinShardLimit is true, it
+	// also tries to select a team whose existing team is less than SERVER_KNOBS->DESIRED_MAX_SHARDS_PER_TEAM.
+	static Optional<Reference<IDataDistributionTeam>> getBestTeam(DDTeamCollection* self,
+	                                                              const GetTeamRequest& req,
+	                                                              bool preferWithinShardLimit) {
+		auto& startIndex = req.preferLowerDiskUtil ? self->lowestUtilizationTeam : self->highestUtilizationTeam;
+		if (startIndex >= self->teams.size()) {
+			startIndex = 0;
+		}
+		Optional<Reference<IDataDistributionTeam>> bestOption;
+		int64_t bestLoadBytes = 0;
+		bool wigglingBestOption = false; // best option contains server in paused wiggle state
+		int bestIndex = startIndex;
+		for (int i = 0; i < self->teams.size(); i++) {
+			int currentIndex = (startIndex + i) % self->teams.size();
+			if (self->teams[currentIndex]->isHealthy()) {
+				int eligibilityType = data_distribution::EligibilityCounter::fromGetTeamRequest(req);
+				if (eligibilityType != data_distribution::EligibilityCounter::NONE &&
+				    self->teams[currentIndex]->getEligibilityCount(eligibilityType) <= 0) {
+					continue;
+				}
+
+				int64_t loadBytes = self->teams[currentIndex]->getLoadBytes(true, req.inflightPenalty);
+				auto team = ShardsAffectedByTeamFailure::Team(self->teams[currentIndex]->getServerIDs(), self->primary);
+				if ((!req.teamMustHaveShards || self->shardsAffectedByTeamFailure->hasShards(team)) &&
+				    // sort conditions
+				    (!bestOption.present() ||
+				     req.lessCompare(bestOption.get(), self->teams[currentIndex], bestLoadBytes, loadBytes))) {
+
+					// bestOption doesn't contain wiggling SS while current team does. Don't replace bestOption
+					// in this case
+					if (bestOption.present() && !wigglingBestOption &&
+					    self->teams[currentIndex]->hasWigglePausedServer()) {
+						continue;
+					}
+
+					if (SERVER_KNOBS->ENFORCE_SHARD_COUNT_PER_TEAM && preferWithinShardLimit &&
+					    self->shardsAffectedByTeamFailure->getNumberOfShards(team) >
+					        SERVER_KNOBS->DESIRED_MAX_SHARDS_PER_TEAM) {
+						continue;
+					}
+
+					bestLoadBytes = loadBytes;
+					bestOption = self->teams[currentIndex];
+					bestIndex = currentIndex;
+					wigglingBestOption = self->teams[bestIndex]->hasWigglePausedServer();
+				}
+			}
+		}
+		startIndex = bestIndex;
```

Comment:
This seems get copied from the original code piece but I am just curious why do we want to do `startIndex = bestIndex;`?

Reply:
startIndex is a referece to either lowestUtilizationTeam or highestUtilizationTeam.

My guess is that it tries to update those two variables at the end of the iteration.